mm/kfence/core.c

```diff
@@ -508,43 +512,47 @@ bool kfence_handle_page_fault(unsigned long addr)
 	lockdep_off();
 	spin_lock_irqsave(&kfence_alloc_lock, flags);
 
-	page_index = (addr - (unsigned long)__kfence_pool) / PAGE_SIZE;
 	if (page_index % 2) {
 		/* This is a redzone, report a buffer overflow. */
-		if (page_index > 1) {
-			obj_index = kfence_addr_to_index(addr - PAGE_SIZE);
-			if (kfence_metadata[obj_index].state == KFENCE_OBJECT_ALLOCATED) {
-				report_index = obj_index;
-				dist = addr - (kfence_metadata[obj_index].addr +
-					       abs(READ_ONCE(kfence_metadata[obj_index].size)));
-			}
-		}
+		struct kfence_metadata *meta = NULL;
+		int distance = 0;
 
-		if (page_index < (CONFIG_KFENCE_NUM_OBJECTS + 1) * 2) {
-			obj_index = kfence_addr_to_index(addr + PAGE_SIZE);
-			if (kfence_metadata[obj_index].state == KFENCE_OBJECT_ALLOCATED) {
-				ndist = kfence_metadata[obj_index].addr - addr;
-				if ((report_index == -1) || (dist > ndist))
-					report_index = obj_index;
-			}
+		meta = kfence_addr_to_metadata(addr - PAGE_SIZE);
+		if (meta && meta->state == KFENCE_OBJECT_ALLOCATED) {
+			to_report = meta;
+			distance = addr - (meta->addr + abs(READ_ONCE(meta->size)));
 		}
 
-		if (report_index == -1) {
-			spin_unlock_irqrestore(&kfence_alloc_lock, flags);
-			pr_err("wild redzone access, possible out-of-bounds access!\n");
-			/* Let the kernel deal with it. */
-			return false;
+		meta = kfence_addr_to_metadata(addr + PAGE_SIZE);
+		if (meta && meta->state == KFENCE_OBJECT_ALLOCATED) {
+			if (!to_report || distance > meta->addr - addr)
+				to_report = meta;
 		}
 
-		kfence_report_error(addr, &kfence_metadata[report_index], KFENCE_ERROR_OOB);
-		kfence_metadata[report_index].unprotected_page = addr;
+		if (!to_report)
+			goto out;
+
+		to_report->unprotected_page = addr;
+		error_type = KFENCE_ERROR_OOB;
 	} else {
-		report_index = kfence_addr_to_index(addr);
+		to_report = kfence_addr_to_metadata(addr);
+		if (!to_report)
+			goto out;
+
 		KFENCE_WARN_ON(!IS_ENABLED(CONFIG_KFENCE_FAULT_INJECTION) &&
-			       kfence_metadata[report_index].state != KFENCE_OBJECT_FREED);
-		kfence_report_error(addr, &kfence_metadata[report_index], KFENCE_ERROR_UAF);
+			       to_report->state != KFENCE_OBJECT_FREED);
+		error_type = KFENCE_ERROR_UAF;
 	}
 
+out:
+	if (to_report)
+		kfence_report_error(addr, to_report, error_type);
+	else
+		/* This may be a UAF or OOB access, but we can't be sure. */
+		// TODO: think about using %p everywhere to not leak kernel
```

Comment:
Ugh, I hate this. For every other tool it's perfectly fine to use %px, as they are test-only.
KFENCE should really avoid printing real addresses on e.g. mobile devices, but I still think it makes sense to print them on syzkaller (should we use KFENCE there) or in server kernels.

Reply:
I fixed this later. It's made conditional on CONFIG_DEBUG_KERNEL.